ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs

```diff
@@ -0,0 +1,321 @@
+
+      -- Note: we don't apply the PRTCL transition, we do that in
+      -- 'updateConsensusState' (called in 'validateHeader') together with
+      -- this function in 'applyExtLedgerState'.
+
+      -- Apply the BBODY transition using the ticked state
+      newShelleyState <- withExcept BBodyError $
+        SL.applyBlockTransition globals tickedShelleyState (shelleyBlockRaw blk)
+
+      let history'
+            -- TODO how expensive is this check?
+            | SL.currentLedgerView oldShelleyState ==
+              SL.currentLedgerView newShelleyState
+            = history
+            | otherwise
+            = History.snapOld
+                (SL.securityParameter globals)
+                slot
+                (SL.currentLedgerView oldShelleyState)
+                history
+
+      return ShelleyLedgerState {
+          ledgerTip    = blockPoint blk
+        , history      = history'
+        , shelleyState = newShelleyState
+        }
+
+  -- TODO actual reapplication:
+  -- https://github.com/input-output-hk/cardano-ledger-specs/issues/1303
+  reapplyLedgerBlock cfg blk ledgerState =
+    case runExcept (applyLedgerBlock cfg blk ledgerState) of
+      Right ledgerState' -> ledgerState'
+      Left  err          -> error $
+        "Reapplication of Shelley ledger block failed: " <> show err
+
+  ledgerTipPoint = ledgerTip
+
+instance TPraosCrypto c => LedgerSupportsProtocol (ShelleyBlock c) where
+  protocolLedgerView _cfg = SL.currentLedgerView . shelleyState
+
+  anachronisticProtocolLedgerView_ cfg ledgerState slot =
+      case History.find slot history of
+        Just lv -> return lv
+        Nothing
+          | slot <  At maxLo
+          -> throwError TooFarBehind -- lower bound is inclusive
+          | slot >= At maxHi
+          -> throwError TooFarAhead  -- upper bound is exclusive
+          | otherwise
+          -> -- If 'futureLedgerView' fails, it is a bug
+             return $ either (error "futureLedgerView failed") id $
+             SL.futureLedgerView globals shelleyState forSlot
+    where
+      ShelleyLedgerState { ledgerTip , history , shelleyState } =
+        ledgerState
+
+      ShelleyLedgerConfig globals = cfg
+
+      k = SL.securityParameter globals
+
+      now, maxHi, maxLo :: SlotNo
+      maxLo = SlotNo $ if (2 * k) > unSlotNo now
+                        then 0
+                        else unSlotNo now - (2 * k)
+      maxHi = SlotNo $ unSlotNo now + (2 * k)
+      -- The slot of the last block last applied to the ledger, /now/
+      -- according to the ledger.
+      now   = fromWithOrigin
+        (minimumPossibleSlotNo (Proxy @(ShelleyBlock c)))
+        (pointSlot ledgerTip)
+
+      -- The slot for which we want to get a 'ProtocolLedgerView'
+      forSlot :: SlotNo
+      forSlot = fromWithOrigin
+        (minimumPossibleSlotNo (Proxy @(ShelleyBlock c)))
+        slot
+
+instance HasHardForkHistory (ShelleyBlock c) where
+  type HardForkIndices (ShelleyBlock c) = '[()]
```

Comment:
The block type of the `HardFork` combinator will deal with this (or a special "Cardano" block that instantiates the `HardFork` combinator to Byron + Shelley . In the future, it will become n-ary and include `Goguen`). `ShelleyBlock` can ignore this and pretend to start from genesis (which we'll do in the tests and the testnet).

The `c` parameter is the crypto, not the config or something similar.